python_asyncio/coroutines_basics_main.py

```diff
@@ -101,50 +103,89 @@ async def coroutine1(msg: str) -> int:
+            class Status(Enum):
+                EXCEPTION = "EXCEPTION"
+                CANCELLED = "CANCELLED"
+                DONE = "DONE"
+                ACTIVE = "-"
+
+                def __init__(self, short_name):
+                    self.short_name = short_name
+
+            status: Status
+            if t.cancelled():
+                status = Status.CANCELLED
+            elif t.done():
+                if t.exception() is not None:
+                    status = Status.EXCEPTION
+                else:
+                    status = Status.DONE
+            else:
+                status = Status.ACTIVE
+            width = max([len(m.short_name) for m in Status.__members__.values()])
+            return status.short_name.center(width)
 
-async def do_something(i: int):
-    logging.info(f"({i}) start")
-    await asyncio.sleep(1)
-    logging.info(f"({i}) end")
+        result: List[str] = []
+        for t in sorted(tasks, key=lambda x: x.get_name()):
+            result.append(f"  [{execution_status(t)}], name={t.get_name()}")
+        return '\n'.join(result)
 
+    # Task is a wrapper around a coroutine that schedules its execution and allows it to run concurrently with other
+    # tasks, managing its lifecycle and state.
+    # Each task has a name, which by default is 'Task-{n}'.
 
-async def with_coroutines():
-    """Create and gather coroutines."""
-    logging.info("Start")
-    c = []
-    for i in range(3):
-        c.append(do_something(i))
+    logging.info("Creating bunch of tasks and forcing them to different states.")
+    tasks: List[asyncio.Task] = []
 
-    # No tasks at this stage, they will be scheduled in gather.
-    log_tasks()
-    logging.info("Sleep")
-    await asyncio.sleep(2)
-    logging.info("Gather")
-    await asyncio.gather(*c)
-    logging.info("End")
+    asyncio.current_task().set_name("Main")
 
+    for i in range(2):
+        tasks.append(asyncio.create_task(coroutine_noop(), name=f"NoOp-{i}"))
 
-async def with_tasks_noawait():
-    """Create tasks and do not await for them to complete."""
-    logging.info("Start")
-    c = []
-    for i in range(3):
-        c.append(asyncio.create_task(do_something(i), name=f"do_something({i})"))
+    t_cancel = asyncio.create_task(coroutine_noop(), name="NoOpToCancel")
+    t_cancel.cancel("Cancelling ad-hoc")
+    tasks.append(t_cancel)
+
+    t_exception = asyncio.create_task(coroutine_raise("Ad-hoc exception"), name="NoOpToRaise")
+    tasks.append(t_exception)
+
+    t_longsleep = asyncio.create_task(coroutine_asleep(60), name="LongAsleep")
+    tasks.append(t_longsleep)
+
+    logging.info(f"Created {len(tasks)} tasks:\n{task_list_pretty_print(tasks)}")
+    # Until tasks are scheduled, nothing happens.
+    assert not t_cancel.done() and not t_cancel.cancelled()
+    assert not t_exception.done()
+    assert not t_longsleep.done()
+
+    await asyncio.sleep(0)
+    # Now all tasks had a chance to be executed, let's see what happened.
```

Comment:
Remove "Now all"
- we do not know if all tasks had a chance or some, I think it is undefined
- - makes it more inpresonal

Reply:
Done.